pkg/ddc/jindofsx/sync_runtime.go

```diff
@@ -0,0 +1,236 @@
+		return fuseChanged, err
+	}
+
+	return
+}
+
+func (e *JindoFSxEngine) syncMasterSpec(ctx cruntime.ReconcileRequestContext, runtime *datav1alpha1.JindoRuntime) (changed bool, err error) {
+	e.Log.V(1).Info("syncMasterSpec")
+	if runtime.Spec.Master.Disabled {
+		return
+	}
+
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		master, err := kubeclient.GetStatefulSet(e.Client, e.getMasterName(), e.namespace)
+		if err != nil {
+			return err
+		}
+
+		if len(runtime.Spec.Master.Resources.Limits) == 0 && len(runtime.Spec.Master.Resources.Requests) == 0 {
+			e.Log.V(1).Info("The resource requirement is not set, skip")
+			return nil
+		}
+
+		masterToUpdate := master.DeepCopy()
+		if len(masterToUpdate.Spec.Template.Spec.Containers) == 1 {
+			masterResources := utils.TranformResourcesWithTieredStore(runtime.Spec.Master.Resources,
+				masterToUpdate.Spec.Template.Spec.Containers[0].Resources)
+			if !utils.ResourceRequirementsEqual(masterToUpdate.Spec.Template.Spec.Containers[0].Resources, masterResources) {
+				e.Log.Info("The resource requirement is different.", "master sts", masterToUpdate.Spec.Template.Spec.Containers[0].Resources,
+					"runtime", masterResources)
+				masterToUpdate.Spec.Template.Spec.Containers[0].Resources =
+					masterResources
+				changed = true
+			} else {
+				e.Log.V(1).Info("The resource requirement of master is the same.")
+			}
+			if changed {
+				if reflect.DeepEqual(master, masterToUpdate) {
+					changed = false
+					e.Log.V(1).Info("The resource requirement of master is not changed, skip")
+					return nil
+				}
+				e.Log.Info("The resource requirement of master is updated")
+				err = e.Client.Update(context.TODO(), masterToUpdate)
+				if err != nil {
+					e.Log.Error(err, "Failed to update the sts spec")
+				}
+			}
+		}
+
+		return err
+	})
+
+	if fluiderrs.IsDeprecated(err) {
+		e.Log.Info("Warning: the current runtime is created by runtime controller before v0.7.0, update specs are not supported. To support these features, please create a new dataset", "details", err)
+		return false, nil
+	}
+
+	return
+}
+
+func (e *JindoFSxEngine) syncWorkerSpec(ctx cruntime.ReconcileRequestContext, runtime *datav1alpha1.JindoRuntime) (changed bool, err error) {
+	e.Log.V(1).Info("syncWorkerSpec")
+	if runtime.Spec.Worker.Disabled {
+		return
+	}
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		workers, err := ctrl.GetWorkersAsStatefulset(e.Client,
+			types.NamespacedName{Namespace: e.namespace, Name: e.getWorkerName()})
+		if err != nil {
+			return err
+		}
+
+		if len(runtime.Spec.Worker.Resources.Limits) == 0 &&
+			len(runtime.Spec.Worker.Resources.Requests) == 0 {
+			e.Log.V(1).Info("The resource requirement of worker is not set, skip")
+			return nil
+		}
+
+		workersToUpdate := workers.DeepCopy()
+		if len(workersToUpdate.Spec.Template.Spec.Containers) == 1 {
+			workerResources := utils.TranformResourcesWithTieredStore(runtime.Spec.Worker.Resources,
```

Comment:
What if users are scaling down the resource requirements of workers.

For example, when a JindoRuntime is created, the runtime spec has a memory request of 10Gi and a tieredstore with 50Gi mem. Then the memory request has been scaled down to 0Gi, which according to the tieredstore mem, it should be synced to 50Gi(the tieredstore mem capacity).

However, it now transforms to the former value(10Gi)

Reply:
Do you mean we should do some validation before downgrading?